copycat/runtime/src/main/java/org/apache/kafka/copycat/runtime/distributed/KafkaConfigStorage.java

```diff
@@ -0,0 +1,438 @@
+                    if (!(newRootConfig instanceof Map)) {
+                        log.error("Invalid data for root config: connectors field should be a Map but is " + newRootConfig.getClass());
+                        return;
+                    }
+                    rootConfig = (Map<String, Integer>) newRootConfig;
+                    updatedTasks = new ArrayList<>(deferredTaskUpdates.keySet());
+                    deferredTaskUpdates.clear();
+                }
+
+                if (!starting)
+                    tasksConfigCallback.onCompletion(null, updatedTasks);
+            } else if (record.key().startsWith(CONNECTOR_PREFIX)) {
+                String connectorName = record.key().substring(CONNECTOR_PREFIX.length());
+                synchronized (lock) {
+                    // Connector configs can be applied and callbacks invoked immediately
+                    if (!(value.value() instanceof Map)) {
+                        log.error("Found connector configuration (" + record.key() + ") in wrong format: " + value.value().getClass());
+                        return;
+                    }
+                    Object newConnectorConfig = ((Map<String, Object>) value.value()).get("properties");
+                    if (!(newConnectorConfig instanceof Map)) {
+                        log.error("Invalid data for connector config: properties filed should be a Map but is " + newConnectorConfig.getClass());
+                        return;
+                    }
+                    connectorConfigs.put(connectorName, (Map<String, String>) newConnectorConfig);
+                }
+                if (!starting)
+                    connectorConfigCallback.onCompletion(null, connectorName);
+            } else if (record.key().startsWith(TASK_PREFIX)) {
+                synchronized (lock) {
+                    ConnectorTaskId taskId = parseTaskId(record.key());
+                    if (taskId == null) {
+                        log.error("Ignoring task configuration because " + record.key() + " couldn't be parsed as a task config key");
+                        return;
+                    }
+                    if (!(value.value() instanceof Map)) {
+                        log.error("Ignoring task configuration because it is in the wrong format: " + value.value());
+                        return;
+                    }
+
+                    Object newTaskConfig = ((Map<String, Object>) value.value()).get("properties");
+                    if (!(newTaskConfig instanceof Map)) {
+                        log.error("Invalid data for task config: properties filed should be a Map but is " + newTaskConfig.getClass());
+                        return;
+                    }
+
+                    deferredTaskUpdates.put(taskId, (Map<String, String>) newTaskConfig);
+                }
+            } else {
+                log.error("Discarding unexpected config update record for key: " + record.key());
```

Comment:
This isn't an inconsistency. It's an invalid key (i.e. none of root, connector, or task config). It indicates somebody wrote invalid data to the topic since the key doesn't match any of the formats we expect. I updated the error message to try to make that clearer.